### ISSUE TITLE
Generate more ergonomic client stubs

### ISSUE THREAD

**Issue by gordonklaus** [improbable-eng/grpc-web#135]
The client stubs generated for all other languages let you write things like `response = client.ServiceMethod(request)`, which is much nicer to use than the `grpc.invoke(method, options)` we have here.

I would expect a unary call to either accept success and failure callbacks or return a Promise.  I suppose streaming calls would have to be callback-only.

**Comment by easyCZ** [comment]
@gordonklaus 

Thanks for your issue. We're currently investigating how to extend the support for this. It may be a while before we modify [ts-protoc-gen](https://github.com/improbable-eng/ts-protoc-gen) to generate stubs like this as we're firstly looking to provide interoperability for `protobuf.js`. 

It would be super useful to see what kind of API would work for you? How would you like to use the client stubs in the case of bi-directional streams? Would you want callback based API or would you rather use something like generators?

**Comment by gordonklaus** [comment]
Generators seem like they would be a good fit for streaming, even bidirectional.  For situations where a loop isn't appropriate, we can just call `await stream.next()` directly.  Errors should be thrown.

**Comment by jonahbron** [comment]
I agree, part of the purpose of RPC is to abstract the fact that a network is involved. 

It's understandable to use the stub generation that's already available for JavaScript and building on top of it. But one thing that confuses me is the necessity to provide network info for every `invoke`. Unless I write my application with another function in between to inject the host, every part of the app needs to know what the host is. 

Is this intentional? Is the solution I described the expected use case?

**Comment by jonahbron** [comment]
I think the API provided by the official grpc-web repository would be really nice to emulate:

https://github.com/grpc/grpc-web/blob/master/net/grpc/gateway/examples/echo/tutorial.md
(mirrored in Pastebin for those without access: https://pastebin.com/riGTp6p7)

Basically something like this for unary

```
var service = new EchoServiceClient('http://localhost:9091');
var unary_request = new EchoRequest();
unary_request.setMessage('foo_unary');
service.Echo(unary_request, {}, function(err, response) {
  console.log('Unary echo response: ' + response.getMessage());
});
```

Or this for server-streaming:

```
var service = new EchoServiceClient('http://localhost:9091');
var stream_request = new ServerStreamingEchoRequest();
var stream = service.ServerStreamingEcho(stream_request);
stream.on('data', function(response) {
  console.log('Server streaming response: '+response.getMessage());
});
```

I think this basic node callback style should be supported at the least.  We could do nice things like promises/generators on top of it too.

**Comment by jonahbron** [comment]
Maybe it would be possible to add the TypeScript support on top of the client-side plugin they've made?

**Comment by jonahbron** [comment]
I've submitted a PR on the protoc plugin to address this issue:

https://github.com/improbable-eng/ts-protoc-gen/pull/40

**Comment by codesuki** [comment]
@easyCZ how is the progress on integrating with `protobuf.js` ? I would like to try that out. Is there any branch for that? 
Our simple test library has almost 400kb, most of it being the compiled proto files. If we cannot get that size down it will become infeasible for us to use grpc. Which would be regrettable.
